### PR TITLE
[Dialect] [Linalgx] Add linalgx ops: 3 vnni matmuls and multi_batch_matmul

### PR DESCRIPTION
**Pull request by LongshengDu**
Added `mm2d_vnni`, `mm4d_vnni`, `batch_reduce_matmul_vnni` with custom verifier, static indexing_maps and iterator_types, and using vnni dims to get constant symbol for indexing_maps.

Added `multi_batch_matmul` with LinalgContractionOpInterface, dynamic indexing_maps and iterator_types.

Tracking: #14

### CODE REVIEW

File: include/gc/Dialect/Linalgx/LinalgxStructuredOps.td [hunk 1]
```diff
@@ -104,4 +104,159 @@ def Linalgx_SigmoidOp : LinalgxStructuredBase_Op<"sigmoid",
   }];
 }
 
+def Linalgx_Mmt2DVnniOp 
+    : LinalgxStructuredBase_Op<"mmt2d_vnni", [AttrSizedOperandSegments]> {
+  let summary = "Transposed matmul with 2d input and vnni packed weights";
+  let description = [{
+    Supported format: A[M, N] * B[N0, K0, k, n, v] -> C[M, K], with:
+      N = N0 * n
+      K = K0 * k * v; v = (1, 2, 4)
```
Comment (ciyongch) [inline include/gc/Dialect/Linalgx/LinalgxStructuredOps.td:113]: Please use M/K for A,  K/N for B, and M/N for C according to convention.
Why do we put `v=1` here?

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp [hunk 2]
```diff
@@ -82,6 +82,385 @@ void SigmoidOp::getEffects(
...
+  maps.back() = simplifyAffineMap(
+      maps.back().replaceDimsAndSymbols({}, symbolBindings, 6, 0));
+  cached = Builder(context).getAffineMapArrayAttr(maps);
+  getOperation()->setAttr(memoizeAttr, cached);
+  return cached;
+}
+
+void Mmt2DVnniOp::regionBuilder(ImplicitLocOpBuilder &b, Block &block,
+                                ArrayRef<NamedAttribute> attrs) {
+  assert(3 > 0 && block.getNumArguments() == 3 &&
```
Comment (ciyongch) [inline lib/gc/Dialect/Linalgx/LinalgxOps.cpp:152]: redundant `3>0`?

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp [hunk 3]
```diff
@@ -82,6 +82,385 @@ void SigmoidOp::getEffects(
...
+  if (!checkRank)
+    return emitOpError() << "not supported input/output shape.";
+  // match M, N, K dims
+  bool matchM = shapeA.getDimSize(0) == shapeC.getDimSize(0);
+  bool matchN =
+      (shapeB.getDimSize(0) * shapeB.getDimSize(3)) == shapeC.getDimSize(1);
+  bool matchK =
+      shapeA.getDimSize(1) ==
+      (shapeB.getDimSize(1) * shapeB.getDimSize(2) * shapeB.getDimSize(4));
+  bool matchVnni = (shapeB.getDimSize(4) == 1) || (shapeB.getDimSize(4) == 2) ||
```
Comment (ciyongch) [inline lib/gc/Dialect/Linalgx/LinalgxOps.cpp:215]: Same as above, why do we consider `1` as vnni format here?

File: test/gc/Dialect/Linlagx/generalize-named-ops.mlir [hunk 4]
```diff
@@ -24,3 +24,84 @@ func.func @generalize_sigmoid(%arg0: tensor<4x256x64xbf16>, %arg1: tensor<4x256x
 // CHECK-NEXT:      linalg.yield %[[DIV]] : bf16
 
 // -----
+
+func.func @generalize_mmt2d_vnni(%arg0: tensor<256x64xf32>, %arg1: tensor<16x2x8x32x4xf32>,
```
Comment (ciyongch) [inline test/gc/Dialect/Linlagx/generalize-named-ops.mlir:28]: Shall we also add failed case for checking?

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp [hunk 5]
```diff
@@ -82,6 +82,385 @@ void SigmoidOp::getEffects(
...
+  if (!result)
+    return emitOpError() << "input/output dims packing not match.";
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// MultiBatchMatmulOp
+//===----------------------------------------------------------------------===//
+
+SmallVector<utils::IteratorType> MultiBatchMatmulOp::getIteratorTypesArray() {
```
Comment (ciyongch) [inline lib/gc/Dialect/Linalgx/LinalgxOps.cpp:364]: Shall we also add `verify` func to check the dims for this op?

File: include/gc/Dialect/Linalgx/LinalgxStructuredOps.td [hunk 6]
```diff
@@ -104,4 +104,159 @@ def Linalgx_SigmoidOp : LinalgxStructuredBase_Op<"sigmoid",
...
+                              Block &block, ArrayRef<NamedAttribute> attrs);
+    static std::function<void(ImplicitLocOpBuilder &,
+                              Block &, ArrayRef<NamedAttribute>)>
+    getRegionBuilder() {
+      return regionBuilder;
+    }
+  }];
+}
+
+def Linalgx_MultiBatchMatmulOp : LinalgxStructuredBase_Op<"multi_batch_matmul",
```
Comment (ZhennanQin) [inline include/gc/Dialect/Linalgx/LinalgxStructuredOps.td:214]: Also define multi_batch_matmul_4d and multi_batch_matmul_4d_vnni?

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp (hunk 3 above)
Comment (LongshengDu) [reply lib/gc/Dialect/Linalgx/LinalgxOps.cpp:215]:
```
input: MKmk/MK
weight: NKkn4k/NKkn2k/NKkn/KN
output: MNmn/MN
```
For NKkn format, can we treat it as NKkn1k so we can reuse this op?

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp (hunk 5 above)
Comment (LongshengDu) [reply lib/gc/Dialect/Linalgx/LinalgxOps.cpp:364]: MultiBatchMatmulOp extends ContractionInterface, which use `mlir::linalg::detail::verifyStructuredOpInterface` and  `mlir::linalg::detail::verifyContractionInterface` to verify. I tested some mismatched dims, it check the dims as expected.

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp (hunk 2 above)
Comment (LongshengDu) [reply lib/gc/Dialect/Linalgx/LinalgxOps.cpp:152]: I reproduced the code structure from `mlir-linalg-ods-yaml-gen` codegen, just for consistency.

File: include/gc/Dialect/Linalgx/LinalgxStructuredOps.td (hunk 6 above)
Comment (LongshengDu) [reply include/gc/Dialect/Linalgx/LinalgxStructuredOps.td:214]: Is this define correct for multi_batch_matmul_4d and multi_batch_matmul_4d_vnni? Do we also want transposed weight?
```
input: BMKmk
weight: BNKkn4k/BNKkn2k/BNKkn1k
output: BMNmn
```

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp (hunk 5 above)
Comment (ciyongch) [reply lib/gc/Dialect/Linalgx/LinalgxOps.cpp:364]: I see, thanks.

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp (hunk 3 above)
Comment (LongshengDu) [reply lib/gc/Dialect/Linalgx/LinalgxOps.cpp:215]: Is there a name for `NKkn` format (differentiating from mmt4d's `Nknk`)?

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp [hunk 7]
```diff
@@ -82,6 +82,385 @@ void SigmoidOp::getEffects(
...
+  static const char memoizeAttr[] = "linalg.memoized_indexing_maps";
+  ArrayAttr cached = getOperation()->getAttrOfType<ArrayAttr>(memoizeAttr);
+  if (cached)
+    return cached;
+
+  static auto mapA = "affine_map<(d0, d1, d2, d3, d4, d5)[s0, s1, s2, s3, "
+                     "s4, s5] -> (d0, (d3 * s4 + d4) * s5 + d5)>";
+  static auto mapB = "affine_map<(d0, d1, d2, d3, d4, d5)[s0, s1, s2, s3, "
+                     "s4, s5] -> (d1, d3, d4, d2, d5)>";
+  static auto mapC = "affine_map<(d0, d1, d2, d3, d4, d5)[s0, s1, s2, s3, "
```
[the remark this answers is not in the document]
Comment (LongshengDu) [reply lib/gc/Dialect/Linalgx/LinalgxOps.cpp:128]: fixed

File: include/gc/Dialect/Linalgx/LinalgxStructuredOps.td (hunk 6 above)
Comment (ZhennanQin) [reply include/gc/Dialect/Linalgx/LinalgxStructuredOps.td:214]: Correct. We don't use transposed weight.

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp (hunk 3 above)
Comment (ZhennanQin) [reply lib/gc/Dialect/Linalgx/LinalgxOps.cpp:215]: > Is there a name for `NKkn` format (differentiating from mmt4d's `Nknk`)?

The name is NKkn. for op naming, I think it's mm4d_vnni(no transpose).

**Comment by zhczhong** [comment]
Do we need to add `batch_reduce_matmul_vnni` so that the `matmul_vnni` could be lowered to `brgemm_vnni` named op?
```
input: BMK
weight: BKN4k/BKN2k
output: MN
```

**Comment by LongshengDu** [comment]
> Do we need to add `batch_reduce_matmul_vnni` so that the `matmul_vnni` could be lowered to `brgemm_vnni` named op?
> 
> ```
> input: BMK
> weight: BKN4k/BKN2k
> output: MN
> ```

Yeah, will add this in the future

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp [hunk 8]
```diff
@@ -82,6 +82,517 @@ void SigmoidOp::getEffects(
...
+  if (!checkRank)
+    return emitOpError() << "not supported input/output shape.";
+  // match M, N, K dims
+  bool matchM = shapeA.getDimSize(0) == shapeC.getDimSize(0);
+  bool matchN =
+      (shapeB.getDimSize(0) * shapeB.getDimSize(3)) == shapeC.getDimSize(1);
+  bool matchK =
+      shapeA.getDimSize(1) ==
+      (shapeB.getDimSize(1) * shapeB.getDimSize(2) * shapeB.getDimSize(4));
+  bool matchVnni = (shapeB.getDimSize(4) == 2) || (shapeB.getDimSize(4) == 4);
```
Comment (yifeizh2) [inline lib/gc/Dialect/Linalgx/LinalgxOps.cpp:215]: Shall we also check vnni dim value based on dtype? (e.g. we restrain vnni to be 2 under bf16; and 4 under u8/s8).

Comment (LongshengDu) [reply lib/gc/Dialect/Linalgx/LinalgxOps.cpp]: Added.

File: lib/gc/Dialect/Linalgx/LinalgxOps.cpp [hunk 9]
```diff
@@ -82,6 +82,531 @@ void SigmoidOp::getEffects(
...
+  return exprs;
+}
+
+ArrayAttr Mm2DVnniOp::getIndexingMaps() {
+  static const char memoizeAttr[] = "linalg.memoized_indexing_maps";
+  ArrayAttr cached = getOperation()->getAttrOfType<ArrayAttr>(memoizeAttr);
+  if (cached)
+    return cached;
+
+  static auto mapA = "affine_map<(d0, d1, d2, d3, d4, d5)[s0, s1, s2, s3, "
```
Comment (ZhennanQin) [inline lib/gc/Dialect/Linalgx/LinalgxOps.cpp:124]: `const char *const` or `constexpr`

**Review by zhczhong — APPROVED** [review]

**Review by ciyongch — APPROVED** [review]

**Pull request merged**